src/page/index.js

```diff
@@ -74,92 +89,88 @@ const api = new Api({
   },
 });
 
-api
-  .getAppInfo()
-  .then(
-    api.getUserInfo().then((res) => {
-      //assign new value for userId
-      userId = res._id;
-      //assign avatar
-      profileAvatar.src = res.avatar;
-      //new UserInfo
-      userInfo = new UserInfo({
-        title: ".profile__title",
-        subtitle: ".profile__subtitle",
-      });
-      //setting userInfo from server
-      userInfo.setUserInfo(res);
-    })
-  )
-  .then(
-    // loaded cards from server
-    api.getInitialCards().then((res) => {
-      //new section
-      cardSection = new Section(
-        {
-          //using data from server
-          data: res,
-          renderer: (item) => {
-            renderCard(item);
-          },
+Promise.all([api.getUserInfo(), api.getInitialCards()])
+  .then(([userData, initialCards]) => {
+    userId = userData._id;
+    userInfo = new UserInfo({
+      title: ".profile__title",
+      subtitle: ".profile__subtitle",
+      avatar: ".profile__picture",
+    });
+    userInfo.setUserInfo(userData);
+    //new Section
+    cardSection = new Section(
+      {
+        //using data from server
+        data: initialCards,
+        renderer: (item) => {
+          renderCard(item);
         },
-        config.cardSectionClass
-      );
-      cardSection.renderItems();
-    })
-  )
+      },
+      config.cardSectionClass
+    );
+    cardSection.renderItems();
+  })
   .catch((err) => {
     console.error(err);
   });
 
 const popupEditForm = new PopupWithForm(
   "#profile__edit-modal",
   (inputValues) => {
-    popupEditForm.toggleSaveBtn();
-    userInfo.setUserInfo(inputValues);
+    popupEditForm.renderLoading(true);
     //updating userInfo (on?at?in) server
     api
       .updateUserInfo(inputValues)
+      .then((res) => {
+        userInfo.setUserInfo(res);
+      })
+      .then(() => {
+        popupEditForm.close();
+      })
       .catch((err) => {
         console.error(err);
       })
       .finally(() => {
-        popupEditForm.close();
-        popupEditForm.toggleSaveBtn();
+        popupEditForm.renderLoading(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```